Purchases/SwiftSources/LocalReceiptParsing/Builders/ASN1ContainerBuilder.swift

```diff
@@ -0,0 +1,80 @@
+//
+// Created by Andrés Boedo on 7/29/20.
+// Copyright (c) 2020 Purchases. All rights reserved.
+//
+
+import Foundation
+
+struct ASN1ContainerBuilder {
+
+    func build(fromPayload payload: ArraySlice<UInt8>) throws -> ASN1Container {
+        guard payload.count >= 2,
+              let firstByte = payload.first else { throw ReceiptReadingError.asn1ParsingError }
+        let containerClass = try extractClass(byte: firstByte)
+        let encodingType = try extractEncodingType(byte: firstByte)
+        let containerType = try extractType(byte: firstByte)
+        let length = try extractLength(data: payload.dropFirst())
+        let identifierTotalBytes = 1
+        let metadataBytes = identifierTotalBytes + length.totalBytes
+        
+        guard payload.count - metadataBytes >= Int(length.value) else { throw ReceiptReadingError.asn1ParsingError }
+        
+        let internalPayload = payload.dropFirst(metadataBytes).prefix(Int(length.value))
+        var internalContainers: [ASN1Container] = []
+        if encodingType == .constructed {
+            var currentPayload = internalPayload
+            while (currentPayload.count > 0) {
+                let internalContainer = try build(fromPayload: currentPayload)
+                internalContainers.append(internalContainer)
```

Comment:
I might kill the recursive creation before the final version, since I'm not sure it's strictly needed.